internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -314,6 +328,35 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 	return cb, nil
 }
 
+func waitForWatcher(ctx context.Context, log *logger.Logger, markerFilePath string, waitTime time.Duration) error {
+	//Wait for the watcher to be up and running
+	watcherContext, cancel := context.WithTimeout(ctx, waitTime)
+	defer cancel()
+
+	markerWatcher := newMarkerFileWatcher(markerFilePath, log)
+	err := markerWatcher.Run(watcherContext)
+	if err != nil {
+		return fmt.Errorf("error starting update marker watcher: %w", err)
+	}
+
+	log.Info("waiting up to %s for upgrade watcher to set %s state in upgrade marker", waitTime, details.StateWatching)
+
+	for {
+		select {
+		case updMarker := <-markerWatcher.Watch():
+			if updMarker.Details != nil && updMarker.Details.State == details.StateWatching {
+				// watcher started and it is watching, all good
+				log.Info("upgrade watcher set %s state in upgrade marker: exiting wait loop", details.StateWatching)
+				return nil
+			}
+
+		case <-watcherContext.Done():
+			log.Error("upgrade watcher did not start watching within %s or context has expired", waitTime)
+			return ErrWatcherNotStarted
```

Comment:
if you capture the error you should get different errors, which will explain if it was a deadline exceeded or a context cancelled.

```suggestion
		case err := <-watcherContext.Done():
			log.Errorf("upgrade watcher did not start watching within %s or context has expired: %w", waitTime, err)
			return ErrWatcherNotStarted
```